alf/algorithms/simple_algorithm.py

```diff
@@ -0,0 +1,228 @@
+# Copyright (c) 2019 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import abstractmethod
+from collections import namedtuple
+
+from tf_agents.policies import tf_policy
+
+LossStep = namedtuple("LossStep", ("trajectory", "critic"))
+
+
+class SimpleAlgorithm(object):
+    """
+    SimpleAgorithm encapsulates the following pattern for many RL algorithms.
+    To follow this pattern, the user need to implement the following functions:
+        predict()
+        pre_critic_step()
+        critic()
+        loss_step()
+        update()
+
+    The following pseudo-code illustrate how these functions are used by
+    SimpleAgent:
+
+    ```python
+    for next_time_step in next_time_steps:
+        critics[i], state = self.pre_critic_step(next_time_step, state)
+    
+    critics = self.critic(next_steps, critics)
+
+    with tr.GradientTape() as tape:
+        for time_step, critic in zip(time_steps, critics):
+            loss_infos[i], state = self.loss_step(time_step, critic)
+
+    total_loss = self._calc_total_loss(loss_infos)
+    
+    grads = tape.gradient(total_loss, variables_to_train)
+
+    self.update(grads)
+    ```
+    """
+
+    def __init__(self, model, policy=None, collect_policy=None,
+                 optimizer=None):
```

Comment:
How do we enable an algorithm to have more than one network, say an actor network and a critic network (in our planner-controller algorithm there would be 3 or 4 networks), should be wrap them into a single model object or let the algorithm to accept multiple models?